src/librustdoc/html/static/js/search.js

```diff
@@ -610,227 +830,257 @@ window.initSearch = function(rawSearchIndex) {
+                        results.totalLev -= lev;
+                    } else {
+                        return true;
+                    }
+                    break;
                 }
-                fullId = ty.id;
-
-                returned = checkReturned(ty, output, true, NO_TYPE_FILTER);
-                if (output.name === "*" || returned) {
-                    in_args = false;
-                    var is_module = false;
+            }
+            if (found) {
+                results.totalLev += lev;
+                results.elems[Math.floor(lev)][fullId] = {
+                    id: id,
+                    index: index,
+                    lev: lev,
+                    dontValidate: isExact,
+                };
+            }
+            return found;
+        }
 
-                    if (input === "*") {
-                        is_module = true;
-                    } else {
-                        var allFound = true;
-                        for (it = 0, len = inputs.length; allFound && it < len; it++) {
-                            allFound = checkType(type, inputs[it], true);
+        function removeUnusedResult(results, lev) {
+            var elems;
+            for (var i = MAX_LEV_DISTANCE; i > lev; --i) {
+                elems = results.elems[i];
+                if (elems.length !== 0) {
+                    for (var entry in elems) {
+                        if (hasOwnPropertyRustdoc(elems, entry)) {
+                            delete elems[entry];
+                            results.totalLev -= lev;
+                            results.nb -= 1;
+                            return true;
                         }
-                        in_args = allFound;
-                    }
-                    if (in_args) {
-                        results_in_args[fullId] = {
-                            id: i,
-                            index: -1,
-                            dontValidate: true,
-                        };
                     }
-                    if (returned) {
-                        results_returned[fullId] = {
-                            id: i,
-                            index: -1,
-                            dontValidate: true,
-                        };
-                    }
-                    if (is_module) {
-                        results[fullId] = {
-                            id: i,
-                            index: -1,
-                            dontValidate: true,
-                        };
+                }
+            }
+            return false;
+        }
+
+        function addIntoResults(isExact, results, fullId, id, index, lev) {
+            if (lev === 0 || (!isExact && lev <= MAX_LEV_DISTANCE)) {
+                if (replaceResult(isExact, results, fullId, id, index, lev)) {
+                    return;
+                } else if (results.nb >= MAX_RESULTS && results.totalLev / results.nb + 0.5 < lev) {
```

Comment:
Where did that 0.5 come from? Is it a heuristic, or is it provably correct in some sense? Are you sure this won’t wind up with results that depend on the order in which things appear in the index file?

Reply:
At this point the item was already "approved". This check is used only when we reached the limit of display (which is 200) for a tab. So in case we have already enough items, and the current `lev` is "too far" from the others, no need to keep it because it won't be displayed afterwards anyway.